### ISSUE TITLE
Get a Storage from an external plugin

### ISSUE THREAD

**Issue by luongnv89** [cyber-geiger/toolbox-communicationApi#7]
**Describe the bug**
Cannot get a storage from an external plugin

**Environmental information**
- Android 8.1.0
- Flutter: 2.8.0
- geiger_api: 0.6.0

**To Reproduce**
- Create 2 different Flutter application, the first one will act as the toolbox core (back-end), and the second one will act as an external plugin
- Minimum version of the first one
```dart
// main.dart
late GeigerApi? localMaster;
late SimpleEventListener? masterListener;
late StorageController? storageController;
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await  initGeigerAPI();
  runApp(MyApp());
}

Future<void> initGeigerAPI() async {
  try {
    flushGeigerApiCache();
    // Init local master
    localMaster =
        await getGeigerApi('', GeigerApi.masterId, Declaration.doNotShareData);
    if (localMaster != null) {
      log('Got the GeigerAPI: ${localMaster.hashCode}');
      await localMaster!.zapState();
      masterListener = SimpleEventListener('master');
      log('masterListener: ${masterListener.hashCode}');
      List<MessageType> allEvents = [MessageType.allEvents];
      await localMaster!.registerListener(allEvents, masterListener!);
      log('Master plugin has been registered and activated: ${masterListener.hashCode}');
      try {
        storageController = localMaster!.getStorage();
        log('Got the StorageController: ${storageController.hashCode}');
      } catch (e2) {
        log('Failed to get the StorageController');
        log(e2.toString());
      }
    } else {
      log('Could not get the GeigerAPI');
    }
  } catch (e) {
    log('Failed to get the GeigerAPI');
    log(e.toString());
  }
}
```

- and the miminum version of the second one
```dart
// main.dart
late GeigerApi? pluginApi;
late StorageController? pluginStorageController;
late SimpleEventListener? pluginListener;
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await initPluginGeigerAPI();
  runApp(MyApp());
}

Future<void> initPluginGeigerAPI() async {
  try {
    flushGeigerApiCache();
    pluginApi = await getGeigerApi(
        './$myPluginId', myPluginId, Declaration.doNotShareData);
    if (pluginApi != null) {
      log('[PLUGIN] Got the GeigerAPI: ${pluginApi.hashCode}');
      try {
        pluginListener = SimpleEventListener('plugin');
        log('[PLUGIN] pluginListener: ${pluginListener.hashCode}');
        List<MessageType> allEvents = [MessageType.allEvents];
        await pluginApi!.registerListener(allEvents, pluginListener!);
        log('[PLUGIN] Plugin has been registered and activated: ${pluginListener.hashCode}');
        try {
          pluginStorageController = pluginApi!.getStorage();
          log('[PLUGIN] Got the StorageController ${pluginStorageController.hashCode}');
        } catch (e3) {
          log('[PLUGIN] Failed to get storage controller');
          log(e3.toString());
        }
      } catch (e2) {
        log('[PLUGIN] Failed to register a plugin');
        log(e2.toString());
      }
    } else {
      log('[PLUGIN] Could not get the GeigerAPI');
    }
  } catch (e) {
    log('[PLUGIN] Failed to get the GeigerAPI');
    log(e.toString());
  }
}
```
**Expected behavior**
- On the log of the first application (the back-end), the geiger api, the storage controller have been initialized successfully.
```sh
[log] Got the GeigerAPI: 138803069
[log] masterListener: 205993890
[log] Master plugin has been registered and activated: 205993890
[log] Got the StorageController: 650854312
```
- On the log of the second application (the external plugin), we can see that the geiger api has been initialized successfully, the plugin has been registered and activated, however we cannot get a storage
```sh
I/flutter (14910): ## got message in plugin __MASTERPLUGIN__ => miCyberrangePluginId=1641551821070>__MASTERPLUGIN__{[REGISTER_PLUGIN] (geiger://__MASTERPLUGIN__/registerPlugin)}
I/flutter (14910): ## notifying PluginListener Eventlistener "master" contains {
I/flutter (14910): }
I/flutter (14910):  for msg REGISTER_PLUGIN geiger://__MASTERPLUGIN__/registerPlugin
I/flutter (14910): ## PluginEvent fired
I/flutter (15181): ## got message in plugin miCyberrangePluginId => __MASTERPLUGIN__=1641551821070>miCyberrangePluginId{[COMAPI_SUCCESS] (geiger://miCyberrangePluginId/registerPlugin)}
I/flutter (15181): ## notifying PluginListener Instance of 'Listener' for msg COMAPI_SUCCESS geiger://miCyberrangePluginId/registerPlugin
I/flutter (15181): ## PluginEvent fired
I/flutter (14910): ## got message in plugin __MASTERPLUGIN__ => miCyberrangePluginId=1641551822562>__MASTERPLUGIN__{[ACTIVATE_PLUGIN] ()}
I/flutter (14910): ## notifying PluginListener Eventlistener "master" contains {
I/flutter (14910):   miCyberrangePluginId=1641551821070>__MASTERPLUGIN__{[REGISTER_PLUGIN] (geiger://__MASTERPLUGIN__/registerPlugin)}
I/flutter (14910): }
I/flutter (14910):  for msg ACTIVATE_PLUGIN null
I/flutter (14910): ## PluginEvent fired
I/flutter (15181): ## got message in plugin miCyberrangePluginId => __MASTERPLUGIN__=1641551822562>miCyberrangePluginId{[COMAPI_SUCCESS] (geiger://miCyberrangePluginId/activatePlugin)}
I/flutter (15181): ## notifying PluginListener Instance of 'Listener' for msg COMAPI_SUCCESS geiger://miCyberrangePluginId/activatePlugin
I/flutter (15181): ## PluginEvent fired
[log] Got the GeigerAPI: 849865144
[log] pluginListener: 414381755
[log] Plugin has been registered and activated: 414381755
[log] Failed to get storage controller
[log] StorageException:Whoops... error while upgrading database (offending: ALTER TABLE storage_node ADD COLUMN last_modified TEXT CHECK( LENGTH(last_modified) <= 20);
      ) at 
...
```
![Screen Shot 2022-01-07 at 11 41 27](https://user-images.githubusercontent.com/3288457/148532205-dfe7ef9f-4e44-4826-9d21-cc51f6db4c3a.png)

log from the first application to show that the plugin has been registered and activated
```
[log] [Listener "master"] received a new event REGISTER_PLUGIN (source: miCyberrangePluginId, target: __MASTERPLUGIN__
[log] Message: miCyberrangePluginId=1641551821070>__MASTERPLUGIN__{[REGISTER_PLUGIN] (geiger://__MASTERPLUGIN__/registerPlugin)}
I/flutter (14910): ## PluginEvent fired
[log] Total events: 1 events
I/flutter (15181): ## got message in plugin miCyberrangePluginId => __MASTERPLUGIN__=1641551821070>miCyberrangePluginId{[COMAPI_SUCCESS] (geiger://miCyberrangePluginId/registerPlugin)}
I/flutter (15181): ## notifying PluginListener Instance of 'Listener' for msg COMAPI_SUCCESS geiger://miCyberrangePluginId/registerPlugin
I/flutter (15181): ## PluginEvent fired
I/flutter (14910): ## got message in plugin __MASTERPLUGIN__ => miCyberrangePluginId=1641551822562>__MASTERPLUGIN__{[ACTIVATE_PLUGIN] ()}
I/flutter (14910): ## notifying PluginListener Eventlistener "master" contains {
I/flutter (14910):   miCyberrangePluginId=1641551821070>__MASTERPLUGIN__{[REGISTER_PLUGIN] (geiger://__MASTERPLUGIN__/registerPlugin)}
I/flutter (14910): }
I/flutter (14910):  for msg ACTIVATE_PLUGIN null
I/flutter (14910): ## PluginEvent fired
[log] [Listener "master"] received a new event ACTIVATE_PLUGIN (source: miCyberrangePluginId, target: __MASTERPLUGIN__
[log] Message: miCyberrangePluginId=1641551822562>__MASTERPLUGIN__{[ACTIVATE_PLUGIN] ()}
[log] Total events: 2 events
I/flutter (15181): ## got message in plugin miCyberrangePluginId => __MASTERPLUGIN__=1641551822562>miCyberrangePluginId{[COMAPI_SUCCESS] (geiger://miCyberrangePluginId/activatePlugin)}
I/flutter (15181): ## notifying PluginListener Instance of 'Listener' for msg COMAPI_SUCCESS geiger://miCyberrangePluginId/activatePlugin
I/flutter (15181): ## PluginEvent fired
```

**Additional context**
Note that if we move the initialization of the GeigerAPI of the external plugin into the first application (means that now both the back-end and the external plugin are in the same application), then we can get a Storage for the external plugin. This explains why the unit test have passed.
```dart
// main.dart
late GeigerApi? localMaster;
late SimpleEventListener? masterListener;
late StorageController? storageController;

late GeigerApi? pluginApi;
late StorageController? pluginStorageController;
late SimpleEventListener? pluginListener;


Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await  initGeigerAPI();
  await initPluginGeigerAPI();
  runApp(MyApp());
}

Future<void> initGeigerAPI() async {
  try {
    flushGeigerApiCache();
    // Init local master
    localMaster =
        await getGeigerApi('', GeigerApi.masterId, Declaration.doNotShareData);
    if (localMaster != null) {
      log('Got the GeigerAPI: ${localMaster.hashCode}');
      await localMaster!.zapState();
      masterListener = SimpleEventListener('master');
      log('masterListener: ${masterListener.hashCode}');
      List<MessageType> allEvents = [MessageType.allEvents];
      await localMaster!.registerListener(allEvents, masterListener!);
      log('Master plugin has been registered and activated: ${masterListener.hashCode}');
      try {
        storageController = localMaster!.getStorage();
        log('Got the StorageController: ${storageController.hashCode}');
      } catch (e2) {
        log('Failed to get the StorageController');
        log(e2.toString());
      }
    } else {
      log('Could not get the GeigerAPI');
    }
  } catch (e) {
    log('Failed to get the GeigerAPI');
    log(e.toString());
  }
}

Future<void> initPluginGeigerAPI() async {
  try {
    flushGeigerApiCache();
    pluginApi = await getGeigerApi(
        './$myPluginId', myPluginId, Declaration.doNotShareData);
    if (pluginApi != null) {
      log('[PLUGIN] Got the GeigerAPI: ${pluginApi.hashCode}');
      try {
        pluginListener = SimpleEventListener('plugin');
        log('[PLUGIN] pluginListener: ${pluginListener.hashCode}');
        List<MessageType> allEvents = [MessageType.allEvents];
        await pluginApi!.registerListener(allEvents, pluginListener!);
        log('[PLUGIN] Plugin has been registered and activated: ${pluginListener.hashCode}');
        try {
          pluginStorageController = pluginApi!.getStorage();
          log('[PLUGIN] Got the StorageController ${pluginStorageController.hashCode}');
        } catch (e3) {
          log('[PLUGIN] Failed to get storage controller');
          log(e3.toString());
        }
      } catch (e2) {
        log('[PLUGIN] Failed to register a plugin');
        log(e2.toString());
      }
    } else {
      log('[PLUGIN] Could not get the GeigerAPI');
    }
  } catch (e) {
    log('[PLUGIN] Failed to get the GeigerAPI');
    log(e.toString());
  }
}
```
- The log shows that the plugin has been registered, activated successfully, and we can get a storage for the plugin.
![Screen Shot 2022-01-07 at 11 45 22](https://user-images.githubusercontent.com/3288457/148532686-b70f0486-6f91-44a4-bf9c-e389ca84ba3a.png)

**Comment by martingwerder** [comment]
This should be addressed in 0.6.2. Can you confirm that it is working now?

**Comment by luongnv89** [comment]
Yeah, I can confirm that this issue has been resolved in 0.6.2. Many thanks to you @martingwerder
